src/worker_pool/balancer/index.ts

```diff
@@ -0,0 +1,43 @@
+import type { PiscinaTask } from '../../task_queue';
+import type { PiscinaWorker } from '..';
+
+export type PiscinaLoadBalancerCommand = {
+  candidate?: PiscinaWorker | null; // If candidate is passed, it will be used as the result of the load balancer and ingore the command
+  command?: 0 | 1; // 1 = add, 0 = busy
+};
+export type PisicnaLoadBalancer = (
+  task: PiscinaTask,
+  workers: PiscinaWorker[]
+) => PiscinaLoadBalancerCommand;
+
+export type ResourceBasedBalancerOptions = {
+  maximumUsage: number;
+};
+export function ResourceBasedBalancer (
+  opts: ResourceBasedBalancerOptions
+): PisicnaLoadBalancer {
+  const { maximumUsage } = opts;
+
+  return (task, workers) => {
+    const command: PiscinaLoadBalancerCommand = { candidate: null, command: 1 };
+    let checkpoint = maximumUsage;
+    for (const worker of workers) {
+      if (worker.currentUsage === 0) {
+        command.candidate = worker;
+        break;
+      }
+
+      if (worker.isRunningAbortableTask) continue;
+
+      if (
+        !task.isAbortable &&
+        (worker.currentUsage < checkpoint)
+      ) {
+        command.candidate = worker;
+        checkpoint = worker.currentUsage;
+      }
+    }
+
+    return command;
```

Comment:
The array index is usually enough. There's race condition left if the worker crash: the LB API shall include a method to handle worker removal. 
The API needs also to include a method to update the LB internal that will be called at the end of each task execution. LB algos might maintain internal statistics to take decision.

Reply:
So reduce the communication semantics to just the _id_ of the worker, so the pool and lb can react to it.

This can be interesting, an `onClose` or `onDestroy` can be called from the pool to hear whenever a worker is down so it can update its internal statistics.

Do you have something specific in mind?

Reply:
An interesting stat is the average tasks runtime  per worker. A custom LB might implement the algo to pick up the worker with the minimum. 
But it will need: 
- per worker histogram for runtime stats
- sensible default stats runtime at worker creation to avoid biased decision in LB algo
The median runtime is also interesting and the LB code will need to compute it since histogram does not compute it. 

Using the `worker_threads` worker id is better than using the array id but at a cost or doing a lookup to find the worker instance. The array index usage has also drawbacks: the code must be very careful at worker removal and  async task handling, the index can stay valid while not pointing to the correct worker. 

Piscina runtime calculation is also lacking some accuracy: it's not exactly the task execution time on the worker, it also includes worker implementation overhead  + message channel latency.

Reply:
The point you made at the latter was one of the reasons I've had some concerns exposing histograms. The time measured is including the message channel overhead + some extra here and there; it will require some tuning to make it fully valuable for something sensitive to precision as an LB; for overall metric gathering can be enough tho.

>Using the worker_threads worker id is better than using the array id but at a cost of doing a lookup to find the worker instance. 

Overall a `Map` might not impose too much overhead, but we can explore; we can start with that and maybe tune for performance afterwards.

Reply:
> The time measured is including the message channel overhead + some extra here and there; it will require some tuning to make it fully valuable for something sensitive to precision as an LB; for overall metric gathering can be enough tho.

Doing accurate runtime measurement will require to change the current worker API: the measurement needs to be done on the worker side. I've not been able to come up with a better solution than extending an existing worker class. 

> 
> > Using the worker_threads worker id is better than using the array id but at a cost of doing a lookup to find the worker instance.
> 
> Overall a `Map` might not impose too much overhead, but we can explore; we can start with that and maybe tune for performance afterwards.

Been there. And finally I've reverted to using an array: 

- `.reduce()` was faster than iteration on `Map` to consolidate any information from the workers such as finding a min. 
- the number of records in the array will stay low, so `Map` will probably remain overkill

Reply:
> Doing accurate runtime measurement will require to change the current worker API: the measurement needs to be done on the worker side. I've not been able to come up with a better solution than extending an existing worker class.

That should be fine, the result comes packed with some extra data from the Worker, we can extend that schema to add that calculation.

> Been there. And finally I've reverted to using an array:

Good catch.
Overall sticking to an array and leaving the LB to decide should be fine, any optimization can come afterwards